### PR TITLE
Garbage collect mesh services on delete and on startup

### PR DESCRIPTION
**Pull request by GrahamGoudeau**
BOT NOTES: 
resolves https://github.com/solo-io/service-mesh-hub/issues/584

### CODE REVIEW

File: services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go [hunk 1]
```diff
@@ -133,44 +235,130 @@ func (m *meshServiceFinder) handleMeshWorkloadUpsert(meshWorkload *zephyr_discov
...
+	allMeshWorkloads, err := m.meshWorkloadClient.ListMeshWorkload(m.ctx, client.MatchingLabels{
+		constants.MESH_PLATFORM: m.clusterName,
+	})
+	if err != nil {
+		logger.Errorf("%+v", err)
+		return err
+	}
+
+	// take care that we don't include the workload that just got deleted
+	// (unsure of the controller-runtime cache behavior here- better safe than sorry)
```
Comment (harveyxia) [inline services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go:287]: @ilackarms If the EventWatcher receives a delete event, is its associated cache guaranteed to _not_ return that deleted object in a `list` or `get`?

File: services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go [hunk 2]
```diff
@@ -133,44 +235,130 @@ func (m *meshServiceFinder) handleMeshWorkloadUpsert(meshWorkload *zephyr_discov
...
+	for _, meshService := range meshServicesOnCluster.Items {
+		kubeService, err := m.serviceClient.GetService(m.ctx, clients.ResourceRefToObjectKey(meshService.Spec.GetKubeService().GetRef()))
+		if errors.IsNotFound(err) {
+			// the service backing this has already disappeared, just clean it up immediately
+			err = m.meshServiceClient.DeleteMeshService(m.ctx, clients.ObjectMetaToObjectKey(meshService.ObjectMeta))
+			if err != nil {
+				return err
+			}
+
+			// we're done for this service, continue to the next one
```
Comment (harveyxia) [inline services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go:329]: nit: the code is clear enough that I don't think we need this comment

File: services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go [hunk 3]
```diff
@@ -62,18 +63,57 @@ func (m *meshServiceFinder) StartDiscovery(
...
+	err = m.garbageCollectOrphanServices(allMeshWorkloads)
+	if err != nil {
+		return err
+	}
 
-	err := serviceEventWatcher.AddEventHandler(m.ctx, &ServiceEventHandler{
-		Ctx:                 m.ctx,
-		HandleServiceUpsert: m.handleServiceUpsert,
+	err = serviceEventWatcher.AddEventHandler(m.ctx, &k8s_core_controller.ServiceEventHandlerFuncs{
+		OnCreate: func(obj *k8s_core_types.Service) error {
```
Comment (harveyxia) [inline services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go:87]: nice, 👍 for using this more elegant pattern

File: services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go [hunk 4]
```diff
@@ -133,44 +235,130 @@ func (m *meshServiceFinder) handleMeshWorkloadUpsert(meshWorkload *zephyr_discov
...
+		return err
+	}
+
+	return nil
+}
+
+// Accepts a list of "live" mesh workloads on the cluster that this mesh service finder is processing
+// This is to support the case where this is run in the context of a delete events rolling in and we want to
+// be careful to avoid considering the one that just got deleted.
+func (m *meshServiceFinder) garbageCollectOrphanServices(liveMeshWorkloads []*zephyr_discovery.MeshWorkload) error {
```
Comment (harveyxia) [inline services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go:310]: I've been using the term "reconcile" to refer to the semantics of checking everything that exists against what SMH has recorded to exist, and deleting the SMH entries that no longer exist

File: services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go [hunk 5]
```diff
@@ -88,41 +128,103 @@ type meshServiceFinder struct {
...
+	if err != nil {
+		logger.Errorf("%+v", err)
+		return nil
+	}
+
+	// not part of any mesh, so no delete action needs to be taken
+	if mesh == nil {
+		return nil
+	}
+	meshService, err := m.buildMeshService(service, mesh, nil, m.clusterName)
```
[the remark this answers is not in the document]
Comment (GrahamGoudeau) [reply services/mesh-discovery/pkg/discovery/mesh-service/k8s/mesh_service_finder.go:169]: Nice, missed that

**Pull request merged**